### PR TITLE
Makefile: standardise and export variables

### PR DESCRIPTION
**Pull request by chrisnovakovic**
This is part of a larger work still in progress (paper deadline looming), but the commits in this PR are good to go now.

This PR makes a number of changes that simplify the makefiles supplied by PRISM:

* The names of the PRISM-specific makefile variables `SRC_DIR`, `CLASSES_DIR`, `OBJ_DIR`, `LIB_DIR`, `INCLUDE_DIR`, `IMAGES_DIR` and `DTDS_DIR` are now prepended with `PRISM_`: they're used across all of the makefiles, and in some cases they're ambiguous unless you look really closely (e.g., lp_solve also uses `SRC_DIR` and `LIB_DIR`, but their values aren't the ones provided by the top-level makefile).
* Rename `C`, `CPP` and `CPPFLAGS` to `CC`, `CXX` and `CXXFLAGS` respectively: [these names are standardised by GNU Make](https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html), and deviating from the standard leads to confusion (I spent longer than I'd like to admit wondering why the C preprocessor was being called so many times :smile:).
* Export all of the above variables in the top-level makefile, eliminating the need for the `$(MAKE) X="$(X)"` incantations when invoking sub-makes and making the interface between PRISM's build chain and the external library build chains more compatible with each other.

I fixed some other minor bugs along the way, mostly by eliminating the need for that code in the first place (e.g., the `make` invocations in the `cuddpackage` and `cuddpackageforce` targets weren't equal, `CXX` being exported in `cuddpackage` but not in `cuddpackageforce`; now `CXX` doesn't need to be specified at all).

Ultimately, this isn't just about eliminating complexity in the makefiles (although I think that's a good thing in its own right): it's currently impossible to specify certain compiler flags in the top-level makefile because of the way sub-makes get invoked, and this cleanup goes part of the way toward solving that problem.

### CODE REVIEW

**Comment by davexparker** [comment]
Looks good. Thanks @chrisnovakovic. We want to get v4.5 pushed out quite imminently, and will look at merging this into the build process after that.

**Comment by davexparker** [comment]
So I tested it on various machines (Mac OS, Ubuntu/Amazon Linux, all 64 bit), without problems.

Cygwin builds currently fail because the new mechanisms for passing CC/CXX do not yet work for the CUDD makefile. It should suffice for now to remove/comment the hard-coded defaults from there. The mechanism for calling CUDD will need to be looked at again (soon) when we move to CUDD 3.0.0.

**Comment by davexparker** [comment]
It would also make sense to fix the current usage of `PRISM_SRC_DIR` (formerly `SRC_DIR`) as part of this PR. There are still (and always have been) instances of `src` being hard-coded, so it is not actually possible at the moment to change the location of the source code via `PRISM_SRC_DIR`.

I think this is will require moving the "Directories" section higher up the Makefile since one occurrence of `src` occurs earlier. The JNI part of the "Directories" section could simultaneously be relocated to the "Java" section.

The same situation likely occurs with `PRISM_LIB_DIR` etc. too.

**Comment by chrisnovakovic** [comment]
> Cygwin builds currently fail because the new mechanisms for passing CC/CXX do not yet work for the CUDD makefile. It should suffice for now to remove/comment the hard-coded defaults from there. The mechanism for calling CUDD will need to be looked at again (soon) when we move to CUDD 3.0.0.

Turns out I'd already fixed this in 550bb87, but forgot to push those later commits to the GitHub branch before opening the PR 🙂

> It would also make sense to fix the current usage of `PRISM_SRC_DIR` (formerly `SRC_DIR`) as part of this PR. There are still (and always have been) instances of `src` being hard-coded, so it is not actually possible at the moment to change the location of the source code via `PRISM_SRC_DIR`.
> 
> The same situation likely occurs with `PRISM_LIB_DIR` etc. too.

Done, for all of them, in a6ddd7a.

Still some more work to do, in particular the automatic detection of Java/JMI-related stuff can probably be simplified.

File: prism/ext/lpsolve55/Makefile [hunk 1]
```diff
@@ -24,26 +24,26 @@ endif
...
+../../$(PRISM_LIB_DIR)/$(LIBPREFIX)lpsolve55$(LIBSUFFIX):
 	mkdir -p lib
 	@(if [ "$(OSTYPE)" = "cygwin" ]; then \
-	  cp bin/$(BINDISTSUFFIX)/$(LIBPREFIX)lpsolve55$(LIBSUFFIX) ../../lib; \
+	  cp bin/$(BINDISTSUFFIX)/$(LIBPREFIX)lpsolve55$(LIBSUFFIX) ../../$(PRISM_LIB_DIR); \
 	else \
-	  (cd src/lp_solve_5.5/lpsolve55 && \
-	  ARCH=$(ARCH) c=$(C) sh -x $(LPSOLVE55_BUILD) \
-	  ) && mv lib/$(LIBPREFIX)lpsolve55$(LIBSUFFIX) ../../lib/; \
+	  (cd $(PRISM_SRC_DIR)/lp_solve_5.5/lpsolve55 && \
```
Comment (davexparker) [inline prism/ext/lpsolve55/Makefile]: `PRISM_SRC_DIR` should stay as `src` here: it refers to the local directory

Comment (chrisnovakovic) [reply prism/ext/lpsolve55/Makefile]: Well-spotted - fixed in 584fb28.

File: prism/Makefile [hunk 2]
```diff
@@ -264,27 +284,18 @@ ifeq ($(OSTYPE),darwin)
 	CLASSPATHSEP = :
 endif
 
-###############
-# Directories #
-###############
+export CFLAGS CXXFLAGS LDFLAGS JFLAGS LIBPREFIX LIBSUFFIX
```
Comment (davexparker) [inline prism/Makefile]: Is there some logic here as to which of the vars defined above this are exported, and which are passed to sub-Makefiles in other ways? There is a problem, currently, that `EXE` is no longer passed down within `make_dirs` (this becomes evident only on Cygwin).

**Comment by davexparker** [comment]
Still fails on Cygwin:
```
/tmp/cc3jj2nK.o:ngprism.c:(.text+0x11): undefined reference to `__imp_WSACleanup'
/tmp/cc3jj2nK.o:ngprism.c:(.text+0x29): undefined reference to `__imp_closesocket'
/tmp/cc3jj2nK.o:ngprism.c:(.text+0xfb): undefined reference to `__imp_send'
...
```
Need to investigate why.

**Comment by davexparker** [comment]
OK, fixing `EXE` solves the Cygwin problem since the variable's existence is tested to include a required shared library on Windows.

File: prism/Makefile (hunk 2 above)
Comment (chrisnovakovic) [reply prism/Makefile]: `EXE` shouldn't be exported (at the moment I'm just exporting standard make variables like compiler paths and flags plus widely-used PRISM-specific variables like `OSTYPE`) - `EXE` was accidentally removed from the `make` invocation in `make_dirs` in 0b21eae9, now added back in c134e7fd.

**Comment by davexparker** [comment]
Thanks @chrisnovakovic. Merged.

**Pull request merged**